src/java/org/apache/cassandra/db/compaction/ShardManagerNodeAware.java

```diff
@@ -123,21 +127,38 @@ private Token[] findTokenAlignedSplitPoints(Token[] sortedTokens, int shardCount
                 // pos is -(insertion point) - 1, so calculate the insertion point
                 pos = -pos - 1;
 
-                // Check the neighbors
-                Token leftNeighbor = sortedTokens[pos - 1];
-                Token rightNeighbor = sortedTokens[pos];
-
-                // Choose the nearest neighbor. By convention, prefer left if value is midpoint.
-                if (value.size(leftNeighbor) <= value.size(rightNeighbor))
+                if (pos == 0)
                 {
-                    nodeAlignedSplitPoints[i] = leftNeighbor;
-                    // No need to bump pos because we decremented it to find the right split token.
+                    // No left neighbor, so choose the right neighbor
+                    nodeAlignedSplitPoints[i] = sortedTokens[pos];
+                    pos++;
                 }
-                else
+                else if (pos == sortedTokens.length)
```

Comment:
This is a potential problem as we may exhaust `sortedTokens` well before we exhaust `evenSplutPoints`.
It looks like we need to set a right-side limit for the binary search as well at `sortedTokens.length - (evenSplitPoints.length - i)` and compare against that here as well.